File: StanceControllerMPC/src/lcm_data_exchange_sc.cpp
#include "lcm_data_exchange_sc.hpp"
#include <unistd.h>

// #define KT 0.74
// #define GEAR_RATIO 10.0

LCMExchanger::LCMExchanger()
{
    
    if(!robot_cmd_subscriber.good())
        return;
    if(!robot_state_subscriber.good())
        return;
    if(!phase_signal_subscriber.good())
        return;
    if(!gait_params_subscriber.good()) 
        return;

    string config_address = mors_sys::GetEnv("CONFIGPATH");//cwd;
    config_address += "/channels.yaml";

    YAML::Node channel_config = YAML::LoadFile(config_address);//"/home/user/mors_mpc_control/config/channels.yaml");
    robot_cmd_channel = channel_config["robot_cmd"].as<string>();
    robot_state_channel = channel_config["robot_state"].as<string>();
    phase_signal_channel = channel_config["gait_phase"].as<string>();
    grf_cmd_channel = channel_config["grf_cmd"].as<string>();
    gait_params_channel = channel_config["gait_params"].as<string>();

    // cmd_vel.resize(6);
    // cmd_pose.resize(6);
    phase.resize(4);
    robot_state.ang_vel.resize(3);
    robot_state.lin_vel.resize(3);
    robot_state.orientation.resize(3);
    robot_state.pos.resize(3);
    robot_cmd.ang_vel.resize(3);
    robot_cmd.lin_vel.resize(3);
    robot_cmd.orientation.resize(3);
    robot_cmd.pos.resize(3);
    leg_state.contacts.resize(4);
    leg_state.r1_pos.resize(3);
    leg_state.l1_pos.resize(3);
    leg_state.r2_pos.resize(3);
    leg_state.l2_pos.resize(3);

    // cmd_vel.setZero();
    // cmd_pose.setZero();
    phase = {1,1,1,1};
    t_gait = 0.0;
    robot_state.ang_vel.setZero();
    robot_state.lin_vel.setZero();
    robot_state.orientation.setZero();
    robot_state.pos.setZero();
    robot_cmd.ang_vel.setZero();
    robot_cmd.lin_vel.setZero();
    robot_cmd.orientation.setZero();
    robot_cmd.pos.setZero();
    leg_state.contacts = {false, false, false, false};
    leg_state.r1_pos.setZero();
    leg_state.l1_pos.setZero();
    leg_state.r2_pos.setZero();
    leg_state.l2_pos.setZero();

    t_sw = 0.0;
    t_st = 0.4;
    gait_type = {0.0, 0, 0, 0.0};
    standing = true;
}

void LCMExchanger::start_exchanger()
{
    thRobotCmd = make_unique<thread> (&LCMExchanger::robotCmdThread, this);
    thRobotState = make_unique<thread> (&LCMExchanger::robotStateThread, this);
    thPhaseSignal = make_unique<thread> (&LCMExchanger::phaseSignalThread, this);
    thGaitParams = make_unique<thread> (&LCMExchanger::gaitParamsThread, this);
}

void LCMExchanger::robotCmdHandler(const lcm::ReceiveBuffer* rbuf,
                                    const std::string& chan,
                                    const mors_msgs::robot_cmd_msg* msg)
{
    for (int i=0; i<3; i++)
    {
        // cmd_pose(i) = msg->cmd_pose[i];
        // cmd_vel(i) = msg->cmd_vel[i];
        robot_cmd.pos(i) = msg->cmd_pose[i];
        robot_cmd.orientation(i) = msg->cmd_pose[i+3];
        robot_cmd.lin_vel(i) = msg->cmd_vel[i];
        robot_cmd.ang_vel(i) = msg->cmd_vel[i+3];
    }
}

void LCMExchanger::robotStateHandler(const lcm::ReceiveBuffer* rbuf,
                                    const std::string& chan,
                                    const mors_msgs::robot_state_msg* msg)
{
    for (int i=0; i<3; i++)
    {
        leg_state.r1_pos(i) = msg->legs.r1_pos[i];
        leg_state.l1_pos(i) = msg->legs.l1_pos[i];
        leg_state.r2_pos(i) = msg->legs.r2_pos[i];
        leg_state.l2_pos(i) = msg->legs.l2_pos[i];

        // leg_state.r1_vel(i) = msg->legs.r1_vel[i];
        // leg_state.l1_vel(i) = msg->legs.l1_vel[i];
        // leg_state.r2_vel(i) = msg->legs.r2_vel[i];
        // leg_state.l2_vel(i) = msg->legs.l2_vel[i];

        // leg_state.r1_grf(i) = msg->legs.r1_grf[i];
        // leg_state.l1_grf(i) = msg->legs.l1_grf[i];
        // leg_state.r2_grf(i) = msg->legs.r2_grf[i];
        // leg_state.l2_grf(i) = msg->legs.l2_grf[i];

        leg_state.contacts[i] = msg->legs.contact_states[i];

        robot_state.pos(i) = msg->body.position[i];
        robot_state.orientation(i) = msg->body.orientation[i];
        robot_state.lin_vel(i) = msg->body.lin_vel[i];
        robot_state.ang_vel(i) = msg->body.ang_vel[i];
    }
    leg_state.contacts[3] = msg->legs.contact_states[3];
}

void LCMExchanger::phaseSignalHandler(const lcm::ReceiveBuffer* rbuf,
                                    const std::string& chan,
                                    const mors_msgs::phase_signal_msg* msg)
{
    // for (int i=0; i<4; i++)
    //     phase(i) = msg->phase[i];

    phase.assign(msg->phase, msg->phase+4);
    phi.assign(msg->phi, msg->phi+4);
    // cout << phase.transpose() << endl;
    t_gait = msg->t;
}

void LCMExchanger::gaitParamsHandler(const lcm::ReceiveBuffer* rbuf,
    const std::string& chan,
    const mors_msgs::gait_params_msg* msg)
{
    // for (int i=0; i<4; i++)
    //     phase(i) = msg->phase[i];
    t_sw = msg->t_sw;
    t_st = msg->t_st;
    standing = msg->standing;
    for (int i = 0; i < 4; i++)
        gait_type[i] = msg->gait_type[i];
    // cout << phase.transpose() << endl;
}

void LCMExchanger::robotCmdThread()
{   
    robot_cmd_subscriber.subscribe(robot_cmd_channel, &LCMExchanger::robotCmdHandler, this);
    while(true)
    {
        robot_cmd_subscriber.handle();
        // cout << "grf thread" << endl;
        // sleep(0.001);
    }
}

void LCMExchanger::robotStateThread()
{   
    robot_state_subscriber.subscribe(robot_state_channel, &LCMExchanger::robotStateHandler, this);
    while(true)
    {
        robot_state_subscriber.handle();
        // cout << "grf thread" << endl;
        // sleep(0.001);
    }
}

void LCMExchanger::phaseSignalThread()
{
    phase_signal_subscriber.subscribe(phase_signal_channel, &LCMExchanger::phaseSignalHandler, this);
    while(true)
    {
        phase_signal_subscriber.handle();
        // cout << "imu thread" << endl;
        // sleep(0.001);
    }
}

void LCMExchanger::gaitParamsThread()
{
    gait_params_subscriber.subscribe(gait_params_channel, &LCMExchanger::gaitParamsHandler, this);
    while(true)
    {
        gait_params_subscriber.handle();
        // cout << "imu thread" << endl;
        // sleep(0.001);
    }
}


void LCMExchanger::sendGrfCmd(LegData &leg_data)
{
    for (int i=0; i<3; i++)
    {
        grfCmdMsg.r1_grf[i] = leg_data.r1_grf(i);
        grfCmdMsg.l1_grf[i] = leg_data.l1_grf(i);
        grfCmdMsg.r2_grf[i] = leg_data.r2_grf(i);
        grfCmdMsg.l2_grf[i] = leg_data.l2_grf(i);
    }
    grf_cmd_publisher.publish(grf_cmd_channel, &grfCmdMsg);
}

RobotData LCMExchanger::getRobotCmd()
{
    return robot_cmd;
}

RobotData LCMExchanger::getBodyState()
{
    return robot_state;
}

LegData LCMExchanger::getLegsState()
{
    return leg_state;
}

vector<int> LCMExchanger::getPhaseSignals()
{
    return phase;
}

double LCMExchanger::getTGait()
{
    return t_gait;
}

vector<double> LCMExchanger::getPhiGait()
{
    return phi;
}

void LCMExchanger::get_gait_params(double& t_st, double& t_sw, vector<double>& gait_type, bool& standing)
{
    t_st = this->t_st;
    t_sw = this->t_sw;
    gait_type = this->gait_type;
    standing = this->standing;
}



LCMExchanger::~LCMExchanger()
{

}

File: SwingController/lcm_data_exchange_sc.py
import lcm
import numpy as np
from mors_msgs.phase_signal_msg import phase_signal_msg
from mors_msgs.robot_state_msg import robot_state_msg
from mors_msgs.robot_cmd_msg import robot_cmd_msg
from mors_msgs.gait_params_msg import gait_params_msg
from mors_msgs.foot_cmd_msg import foot_cmd_msg

R1 = 0
L1 = 1
R2 = 2
L2 = 3

class LCMDataExchange():
    def __init__(self,
                 gait_params_cnannel = "GAIT_PARAMS",
                 gait_phase_channel = "GAIT_PHASE",
                 robot_state_channel = "ROBOT_STATE",
                 robot_cmd_channel = "ROBOT_CMD",
                 foot_cmd_channel = "FOOT_CMD",
                 ) -> None:
        self.gait_params_cnannel = gait_params_cnannel
        self.gait_phase_channel = gait_phase_channel
        self.robot_state_channel = robot_state_channel
        self.robot_cmd_channel = robot_cmd_channel
        self.foot_cmd_channel = foot_cmd_channel

        self.t_st = 0.4
        self.t_sw = 0.25
        self.stride_height = 0.05

        self.gait_phase = [0]*4
        self.gait_phi = [0]*4

        self.ref_body_height = 0.0
        self.ref_body_yaw_vel = 0.0
        self.ref_body_vel = [0.0, 0.0, 0.0]

        self.base_pos = [0.0]*3
        self.base_orientation = [0.0]*3
        self.base_ang_vel = [0.0]*3
        self.base_lin_vel = [0.0]*3

        self.r1_pos = [0.0]*3
        self.l1_pos = [0.0]*3
        self.r2_pos = [0.0]*3
        self.l2_pos = [0.0]*3

        self.phase_sig_msg = phase_signal_msg()
        self.rbt_state_msg = robot_state_msg()
        self.rbt_cmd_msg = robot_cmd_msg()
        self.gait_prms_msg = gait_params_msg()
        self.foot_cmd_msg = foot_cmd_msg()
        self.lc_cmd = lcm.LCM()
        

    def robot_state_callback(self, channel, data : robot_state_msg):
        state_msg = robot_state_msg.decode(data)
        self.base_pos = state_msg.body.position[:]
        self.base_lin_vel = state_msg.body.lin_vel[:]
        self.base_orientation = state_msg.body.orientation[:]
        self.base_ang_vel = state_msg.body.ang_vel[:]
        self.r1_pos = state_msg.legs.r1_pos[:]
        self.l1_pos = state_msg.legs.l1_pos[:]
        self.r2_pos = state_msg.legs.r2_pos[:]
        self.l2_pos = state_msg.legs.l2_pos[:]


    def robot_state_thread(self):
        print("RobotState thread started")
        lc = lcm.LCM()
        subscription = lc.subscribe(self.robot_state_channel, self.robot_state_callback)
        try:
            while True:
                lc.handle()
        except KeyboardInterrupt:
            pass

    def robot_cmd_callback(self, channel, data : robot_cmd_msg):
        msg = robot_cmd_msg.decode(data)
        self.ref_body_height = msg.cmd_pose[2]
        self.ref_body_yaw_vel = msg.cmd_vel[5]
        self.ref_body_vel = msg.cmd_vel[0:3]


    def robot_cmd_thread(self):
        print("RobotCmd thread started")
        lc = lcm.LCM()
        subscription = lc.subscribe(self.robot_cmd_channel, self.robot_cmd_callback)
        try:
            while True:
                lc.handle()
        except KeyboardInterrupt:
            pass

    def gait_params_callback(self, channel, data : gait_params_msg):
        msg = gait_params_msg.decode(data)
        self.t_st = msg.t_st
        self.t_sw = msg.t_sw
        self.stride_height = msg.stride_height

    def gait_params_thread(self):
        print("GaitParams thread started")
        lc = lcm.LCM()
        subscription = lc.subscribe(self.gait_params_cnannel, self.gait_params_callback)
        try:
            while True:
                lc.handle()
        except KeyboardInterrupt:
            pass
    
    def gait_phase_callback(self, channel, data : phase_signal_msg):
        msg = phase_signal_msg.decode(data)
        self.gait_phase = msg.phase[:]
        self.gait_phi = msg.phi

    def gait_phase_thread(self):
        print("GaitPhase thread started")
        lc = lcm.LCM()
        subscription = lc.subscribe(self.gait_phase_channel, self.gait_phase_callback)
        try:
            while True:
                lc.handle()
        except KeyboardInterrupt:
            pass

    
    def get_gait_params(self):
        return self.t_st, self.t_sw, self.stride_height
    
    def get_gait_phase(self):
        return self.gait_phase, self.gait_phi
    
    def get_body_state(self):
        return self.base_pos, self.base_orientation, self.base_lin_vel, self.base_ang_vel
    
    def get_legs_pos(self):
        return self.r1_pos, self.l1_pos, self.r2_pos, self.l2_pos
    
    def get_robot_cmd(self):
        return self.ref_body_height, self.ref_body_yaw_vel, self.ref_body_vel
    
    def set_kpkd(self, kp, kd):
        self.foot_cmd_msg.Kp = kp[:]
        self.foot_cmd_msg.Kd = kd[:]

    def send_foot_cmd(self, x, dx, ddx):
        self.foot_cmd_msg.r1_pos = x[R1][:]
        self.foot_cmd_msg.l1_pos = x[L1][:]
        self.foot_cmd_msg.r2_pos = x[R2][:]
        self.foot_cmd_msg.l2_pos = x[L2][:]

        self.foot_cmd_msg.r1_vel = dx[R1][:]
        self.foot_cmd_msg.l1_vel = dx[L1][:]
        self.foot_cmd_msg.r2_vel = dx[R2][:]
        self.foot_cmd_msg.l2_vel = dx[L2][:]

        self.foot_cmd_msg.r1_acc = ddx[R1][:]
        self.foot_cmd_msg.l1_acc = ddx[L1][:]
        self.foot_cmd_msg.r2_acc = ddx[R2][:]
        self.foot_cmd_msg.l2_acc = ddx[L2][:]

        self.lc_cmd.publish(self.foot_cmd_channel, self.foot_cmd_msg.encode())